src/dex/swaap-v1/swaap-v1.ts

```diff
@@ -0,0 +1,633 @@
+  protected exchangeProxy: string;
+
+  protected eventPools: { [poolAddress: string]: SwaapV1Pool } = {};
+
+  public allPools: Map<string, SubgraphPoolBase> = new Map();
+
+  readonly hasConstantPriceLargeAmounts = false;
+  readonly isFeeOnTransferSupported: boolean = false;
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(SwaapV1Config);
+
+  logger: Logger;
+
+  constructor(
+    protected network: Network,
+    protected dexKey: string,
+    protected dexHelper: IDexHelper,
+    protected adapters = Adapters[network], // TODO: add any additional optional params to support other fork DEXes
+  ) {
+    super(dexHelper.config.data.augustusAddress, dexHelper.web3Provider);
+    this.config = SwaapV1Config[dexKey][network];
+    this.logger = dexHelper.getLogger(`${dexKey}-${network}`);
+    this.subgraphURL = this.config.subgraphURL;
+    this.exchangeProxy = this.config.exchangeProxy;
+  }
+
+  async init(blockNumber: number) {
+    this.allPools = await this.getAllPools(blockNumber);
+  }
+
+  // Initialize pricing is called once in the start of
+  // pricing service. It is intended to setup the integration
+  // for pricing requests. It is optional for a DEX to
+  // implement this function
+  async initializePricing(blockNumber: number) {
+    await this.init(blockNumber);
+    if (!this.allPools)
+      throw new Error(
+        'initializePricing: SwaapV1 cfgInfo still null after init',
+      );
+    for (const poolConfig of this.allPools.values()) {
+      await this.addPool(poolConfig, blockNumber);
+    }
+  }
+
+  async addPool(poolConfig: SubgraphPoolBase, blockNumber: number) {
+    const pool = new SwaapV1Pool(
+      this.dexKey,
+      this.network,
+      poolConfig,
+      this.dexHelper,
+    );
+    const state = await pool.generateState(blockNumber);
+    pool.setState(state, blockNumber);
+
+    this.dexHelper.blockManager.subscribeToLogs(
+      pool,
+      pool.addressSubscribers,
+      blockNumber,
+    );
+    this.eventPools[poolConfig.id] = pool;
+  }
+
+  // Returns estimated gas cost of calldata for this DEX in multiSwap
+  getCalldataGasCost(poolPrices: PoolPrices<SwaapV1Data>): number | number[] {
+    // TODO: update if there is any payload in getAdapterParam
+    return CALLDATA_GAS_COST.DEX_NO_PAYLOAD;
+  }
+
+  // Returns the list of contract adapters (name and index)
+  // for a buy/sell. Return null if there are no adapters.
+  getAdapters(side: SwapSide): { name: string; index: number }[] | null {
+    return this.adapters[side];
+  }
+
+  protected findPools(
+    srcTokenAddress: Address,
+    destTokenAddress: Address,
+  ): Address[] {
+    if (!this.allPools) return [];
+    return Array.from(this.allPools.values())
+      .filter(
+        poolConfig =>
+          poolConfig.tokens.map(t => t.address).includes(srcTokenAddress) &&
+          poolConfig.tokens.map(t => t.address).includes(destTokenAddress),
+      )
```

Comment:
This is not really efficient if we have lot of pools. Maybe it can be improve ?

Reply:
Using another data structure than an array?

Reply:
KyberDMM impl of the findPools is quite similar (actually we copied their impl).
Is it fine if we add an intermediary variable like this?
```.filter(
        poolConfig => {
            const tokenAddresses = poolConfig.tokens.map(t => t.address)
            return tokenAddresses.includes(srcTokenAddress) && tokenAddresses.includes(destTokenAddress)
          }
        ,
      )
```

Reply:
ok thanks!